src/Drupal/Commands/core/MigrateRunnerCommands.php

```diff
@@ -0,0 +1,706 @@
+     *   level: Level
+     *   message: Message
+     *   hash: Source IDs hash
+     * @default-fields level,message,hash
+     *
+     * @return \Consolidation\OutputFormatters\StructuredData\RowsOfFields
+     *   Migration messages status formatted as table.
+     */
+    public function messages(string $migrationId): RowsOfFields
+    {
+        /** @var \Drupal\migrate\Plugin\MigrationInterface $migration */
+        $migration = $this->getMigrationPluginManager()->createInstance($migrationId);
+        $table = [];
+        foreach ($migration->getIdMap()->getMessages() as $row) {
+            $table[] = [
+                'level' => $row->level,
+                'message' => $row->message,
+                'hash' => $row->source_ids_hash,
+            ];
+        }
+        return new RowsOfFields($table);
+    }
+
+    /**
+     * List the fields available for mapping in a source.
+     *
+     * @command migrate:fields-source
+     *
+     * @drupal-dependencies migrate
+     *
+     * @param string $migrationId
+     *   The ID of the migration.
+     *
+     * @usage migrate:fields-source article
+     *   List fields for the source in the article migration
+     *
+     * @aliases mfs,migrate-fields-source
+     *
+     * @topics docs:migrate
+     *
+     * @validate-module-enabled migrate
+     *
+     * @field-labels
+     *   machine_name: Field name
+     *   description: Description
+     * @default-fields machine_name,description
+     *
+     * @return \Consolidation\OutputFormatters\StructuredData\RowsOfFields
+     *   Migration messages status formatted as table.
+     */
+    public function fieldsSource(string $migrationId): RowsOfFields
+    {
+        /** @var \Drupal\migrate\Plugin\MigrationInterface $migration */
+        $migration = $this->getMigrationPluginManager()->createInstance($migrationId);
+        $source = $migration->getSourcePlugin();
+        $table = [];
+        foreach ($source->fields() as $machineName => $description) {
+            $table[] = [
+                'machine_name' => $machineName,
+                'description' => strip_tags($description),
+            ];
+        }
+        return new RowsOfFields($table);
+    }
+
+    /**
+     * Retrieves a list of active migrations.
+     *
+     * @param string|null $migrationIds A comma-separated list of migration IDs. If omitted, will return all
+     *   migrations.
+     * @param string|null $tags
+     *   A comma separated list of tags.
+     *
+     * @return \Drupal\migrate\Plugin\MigrationInterface[][]
+     *   An array keyed by migration tag, each value containing an array of
+     *   migrations or an empty array if no migrations match the input criteria.
+     *
+     * @throws \Drupal\Component\Plugin\Exception\PluginException
+     */
+    protected function getMigrationList(?string $migrationIds, ?string $tags): array
+    {
+        $migrationIds = array_filter(array_map('trim', explode(',', $migrationIds)));
```

Comment:
Perhaps use `\Drush\Utils\StringUtils::csvToArray` here.

Reply:
I think it works. However, `StringUtils::csvToArray()` might need an outer `array_values()` as the result shows index gaps on consecutive commas.